example/notify_inval_entry.c

```diff
@@ -260,6 +271,10 @@ static void* update_fs_loop(void *data) {
             if(options.only_expire) {
                 assert(fuse_lowlevel_notify_expire_entry
                    (se, FUSE_ROOT_ID, old_name, strlen(old_name), FUSE_LL_EXPIRE_ONLY) == 0);
+                // For some reason, the kernel doesn't send a forget request,
+                // so we have to do it ourselves.
+                // (At least that's the behaviour on 6.2.9)
+                lookup_cnt = 0;
```

Comment:
Kernel will send a forget request once it gives up the corresponding inode. So now the invalidate request sends FUSE_EXPIRE_ONLY, which basically suppresses d_invalidate() on the kernel side.

`fuse_reverse_inval_entry()`
```

	fuse_dir_changed(parent);
	if (!(flags & FUSE_EXPIRE_ONLY))
		d_invalidate(entry);
	fuse_invalidate_entry_cache(entry);

	if (child_nodeid != 0 && d_really_is_positive(entry)) {
```
The caller (`fuse_notify_inval_entry`) does not set `child_nodeid`, so we can ignore that code path - the request just invalidates the fuse cache. On _next_ file access (stat/open/whatever) vfs will trigger a revalidate and if that fails dentry will be removed. Once the last inode reference is dropped a forget will be eventually send.

Regarding the test - without file access in between, lookup is not dropped - easiest way to get that would be something like `stat(old_name, ...)`.
Also, lookup_cnt is not lock protected - if anything would run in parallel - test will give unexpected results.

Reply:
And while we are at it, looking at kernel code, I'm worried about a missing test if daemon side supports flags at all - it seems to rely on that the buffer was zeroed in previous lib versions. Now that is implementation dependent - what guarantees zeroing of a padded value?

Reply:
> Also, lookup_cnt is not lock protected - if anything would run in parallel - test will give unexpected results.

Yes, that's already a problem with the existing code, I think?  Might be worth a separate PR.

(I'm still digesting the rest of what you explained.)

Reply:
> And while we are at it, looking at kernel code, I'm worried about a missing test if daemon side supports flags at all - it seems to rely on that the buffer was zeroed in previous lib versions. Now that is implementation dependent - what guarantees zeroing of a padded value?

Depending on exactly how you allocate your struct, in most cases using uninitialized variables (or buffers) in C is much worse than just getting the occasional unexpected value: it's undefined behaviour, and your program might do anything; including formatting your hard drive or sending embarrassing emails.

Luckily, that also means that tools like Valgrind are on the lookout for such omissions.  It seems like the commit 24b35c3d97ffdbf0a1f8e8b4e94ed892343603a6 that introduces `fuse_lowlevel_notify_inval_entry` already has `outarg.padding = 0;`.

You can see the evolution of that function with `git log -L :fuse_lowlevel_notify_inval_entry:lib/fuse_lowlevel.c`.

If you run `git log --pickaxe-regex -S 'padding *= *0' --stat --patch`, you can see a few more instances where valgrind lead to more careful coding, like c2a47b08afdfb75bbdd7dec62336bd0a04346628 or df0d539d04247dbbfcb10c364638df7edaef697e.

Reply:
> [...] Once the last inode reference is dropped a forget will be eventually send.

Ok, it seems to properly exercise this functionality, we should probably have more than one inode?

Reply:
> > And while we are at it, looking at kernel code, I'm worried about a missing test if daemon side supports flags at all - it seems to rely on that the buffer was zeroed in previous lib versions. Now that is implementation dependent - what guarantees zeroing of a padded value?
> 
> Depending on exactly how you allocate your struct, in most cases using uninitialized variables (or buffers) in C is much worse than just getting the occasional unexpected value: it's undefined behaviour, and your program might do anything; including formatting your hard drive or sending embarrassing emails.
> 
> Luckily, that also means that tools like Valgrind are on the lookout for such omissions. It seems like the commit [24b35c3](https://github.com/libfuse/libfuse/commit/24b35c3d97ffdbf0a1f8e8b4e94ed892343603a6) that introduces `fuse_lowlevel_notify_inval_entry` already has `outarg.padding = 0;`. That commit's from 2009.
> 
> You can see the evolution of that function with `git log -L :fuse_lowlevel_notify_inval_entry:lib/fuse_lowlevel.c`. (EDIT: Or rather `git log -L :fuse_lowlevel_notify_expire_entry:lib/fuse_lowlevel.c` because of a rename.)
> 
> If you run `git log --pickaxe-regex -S 'padding *= *0' --stat --patch`, you can see a few more instances where valgrind lead to more careful coding, like [c2a47b0](https://github.com/libfuse/libfuse/commit/c2a47b08afdfb75bbdd7dec62336bd0a04346628) or [df0d539](https://github.com/libfuse/libfuse/commit/df0d539d04247dbbfcb10c364638df7edaef697e).

That is libfuse, what about all other (including possibly unknown) fuse users? The kernel code assumes the buffer is zeroed - I don't think that requirement is documented anywhere, so implementation depending... Reminder to myself, I need to be more active with fuse kernel patch review ;) This already the 2nd instance of such an issue. 
https://patchwork.kernel.org/project/linux-fsdevel/patch/165002363635.1457422.5930635235733982079.stgit@localhost/#24825548